libnymea-core/devices/scriptdeviceplugin.cpp

```diff
@@ -0,0 +1,255 @@
+ *                                                                         *
+ *  You should have received a copy of the GNU General Public License      *
+ *  along with nymea. If not, see <http://www.gnu.org/licenses/>.          *
+ *                                                                         *
+ * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * */
+
+#include "scriptdeviceplugin.h"
+
+#include <QQmlEngine>
+#include <QDir>
+#include <QJsonDocument>
+
+#include "loggingcategories.h"
+#include <plugintimer.h>
+
+ScriptDevicePlugin::ScriptDevicePlugin(QObject *parent) : DevicePlugin(parent)
+{
+
+}
+
+bool ScriptDevicePlugin::loadScript(const QString &fileName)
+{
+
+    QFileInfo fi(fileName);
+    QString metaDataFileName = fi.absoluteDir().path() + '/' + fi.baseName() + ".json";
+
+    QFile metaDataFile(metaDataFileName);
+    if (!metaDataFile.open(QFile::ReadOnly)) {
+        qCWarning(dcDeviceManager()) << "Failed to open plugin metadata at" << metaDataFileName;
+        return false;
+    }
+    QJsonParseError error;
+    QByteArray data = metaDataFile.readAll();
+    QJsonDocument jsonDoc = QJsonDocument::fromJson(data, &error);
+    metaDataFile.close();
+
+    if (error.error != QJsonParseError::NoError) {
+        int errorOffset = error.offset;
+        int newLineIndex = data.indexOf("\n");
+        int lineIndex = 1;
+        while (newLineIndex > 0 && errorOffset > newLineIndex) {
+            data.remove(0, newLineIndex + 2);
+            errorOffset -= (newLineIndex + 2);
+            newLineIndex = data.indexOf("\n");
+            lineIndex++;
+        }
+        if (newLineIndex >= 0) {
+            data = data.left(newLineIndex);
+        }
+        QString spacer;
+        for (int i = 0; i < errorOffset; i++) {
+            spacer += ' ';
+        }
+        QDebug dbg = qWarning(dcDeviceManager()).nospace().noquote();
+        dbg << metaDataFileName << ":" << lineIndex << ":" << errorOffset + 2 << ": error: JSON parsing failed: " << error.errorString() << ": " << data.trimmed() << endl;
+        dbg << data << endl;
+        dbg << spacer << "^";
+        return false;
+    }
+    m_metaData = QJsonObject::fromVariantMap(jsonDoc.toVariant().toMap());
+
+    m_engine = new QQmlEngine(this);
+    m_engine->installExtensions(QJSEngine::AllExtensions);
+
+    QJSValue deviceMetaObject = m_engine->newQMetaObject(&Device::staticMetaObject);
+    m_engine->globalObject().setProperty("Device", deviceMetaObject);
+
+    m_pluginImport = m_engine->importModule(fileName);
+    if (m_pluginImport.isError()) {
+        qCWarning(dcDeviceManager()) << "Error loading plugin module" << m_pluginImport.errorType() << m_pluginImport.toString();
+        return false;
+    }
+
+    return true;
+}
+
+QJsonObject ScriptDevicePlugin::metaData() const
+{
+    return m_metaData;
+}
+
+void ScriptDevicePlugin::init()
+{
+    qmlRegisterType<PluginTimerManager>();
+    qmlRegisterType<PluginTimer>();
```

Comment:
The variant you posted is not valid... there are only
`int qmlRegisterType()`
and
`int qmlRegisterType(const char *uri, int versionMajor, int versionMinor, const char *qmlName)`

The difference being that the former registers the type anonymously, while the second would allow qml code to create instances of it. However, we're not even running full qml in there (i.e. a plugin never does "import nymea 1.0" and also we don't want plugins to create a new PluginTimer just by "new PluginTimer()"...

So this is actually what we want...

With Qt 5.14 this is going to be deprecated and `qmlRegisterAnonymousType()` is going to replace it, but for now we'll have to stick to this.